controller/src/settings/esp.rs

```diff
@@ -88,47 +102,73 @@ impl EspColor {
         }
     }
 
+    fn interpolate_color(start: [f32; 4], end: [f32; 4], t: f32) -> [f32; 4] {
+        [
+            start[0] + (end[0] - start[0]) * t,
+            start[1] + (end[1] - start[1]) * t,
+            start[2] + (end[2] - start[2]) * t,
+            start[3] + (end[3] - start[3]) * t,
+        ]
+    }
+
     /// Calculate the target color.
     /// Health should be in [0.0;1.0]
     pub fn calculate_color(&self, health: f32, distance: f32) -> [f32; 4] {
         match self {
             Self::Static { value } => value.as_f32(),
-            Self::HealthBased { max, min } => {
-                let min_rgb = min.as_f32();
+            Self::HealthBased {
+                max,
+                mid,
+                min,
+                alpha,
+            } => {
                 let max_rgb = max.as_f32();
+                let mid_rgb = mid.as_f32();
+                let min_rgb = min.as_f32();
 
-                [
-                    min_rgb[0] + (max_rgb[0] - min_rgb[0]) * health,
-                    min_rgb[1] + (max_rgb[1] - min_rgb[1]) * health,
-                    min_rgb[2] + (max_rgb[2] - min_rgb[2]) * health,
-                    min_rgb[3] + (max_rgb[3] - min_rgb[3]) * health,
-                ]
+                let color = if health > 0.5 {
+                    let t = (health - 0.5) * 2.0;
+                    Self::interpolate_color(mid_rgb, max_rgb, t)
+                } else {
+                    let t = health * 2.0;
+                    Self::interpolate_color(min_rgb, mid_rgb, t)
+                };
+                [color[0], color[1], color[2], *alpha]
```

Comment:
No need for this. You can just register the result of the if statement (as colour already contains an alpha component)